pkg/envtest/apiserver.go

```diff
@@ -0,0 +1,443 @@
+	"github.com/gardener/gardener/pkg/utils/secrets"
+)
+
+const (
+	envGardenerAPIServerBin = "TEST_ASSET_GARDENER_APISERVER"
+	waitPollInterval        = 100 * time.Millisecond
+)
+
+// GardenerAPIServer knows how to start, register and stop a temporary gardener-apiserver instance.
+type GardenerAPIServer struct {
+	// EtcdURL is the etcd URL that the APIServer should connect to (defaults to the URL of the envtest etcd).
+	EtcdURL *url.URL
+	// CertDir is a path to a directory containing whatever certificates the APIServer needs.
+	// If left unspecified, then Start will create a temporary directory and generate the needed
+	// certs and Stop will clean it up.
+	CertDir string
+	// caCert is the certificate of the CA that signed the GardenerAPIServer's serving cert.
+	caCert *secrets.Certificate
+	// restConfig is used to setup and register the APIServer with the envtest kube-apiserver.
+	restConfig *rest.Config
+	// Path is the path to the gardener-apiserver binary, can be set via TEST_ASSET_GARDENER_APISERVER.
+	// If Path is unset, gardener-apiserver will be started in-process.
+	Path string
+	// SecurePort is the secure port that the APIServer should listen on.
+	// If this is not specified, we default to a random free port on localhost.
+	SecurePort int
+	// listenURL is the URL we end up listening on.
+	listenURL *url.URL
+	// Args is a list of arguments which will passed to the APIServer binary.
+	// If not specified, the minimal set of arguments to run the APIServer will
+	// be used.
+	Args []string
+	// StartTimeout, StopTimeout specify the time the APIServer is allowed to
+	// take when starting and stoppping before an error is emitted.
+	// If not specified, these default to 20 seconds.
+	StartTimeout time.Duration
+	StopTimeout  time.Duration
+	// Out, Err specify where APIServer should write its StdOut, StdErr to.
+	// If not specified, the output will be discarded.
+	Out io.Writer
+	Err io.Writer
+	// HealthCheckEndpoint is the path of the healthcheck endpoint (defaults to "/healthz").
+	// It will be polled until receiving http.StatusOK (or StartTimeout occurs), before
+	// returning from Start.
+	HealthCheckEndpoint string
+
+	// terminateFunc holds a func that will terminate this GardenerAPIServer.
+	terminateFunc func()
+	// exited is a channel that will be closed, when this GardenerAPIServer exits.
+	exited chan struct{}
+}
+
+// Start brings up the GardenerAPIServer, waits for it to be healthy and registers Gardener's APIs.
+func (a *GardenerAPIServer) Start() error {
+	if err := a.defaultSettings(); err != nil {
+		return err
+	}
+
+	a.exited = make(chan struct{})
+	if a.Path != "" {
+		if err := a.runAPIServerBinary(); err != nil {
+			return err
+		}
+	} else {
+		if err := a.runAPIServerInProcess(); err != nil {
+			return err
+		}
+	}
+
+	startCtx, cancel := context.WithTimeout(context.Background(), a.StartTimeout)
+	defer cancel()
+
+	// TODO: retry starting GardenerAPIServer on failure
```

Comment:
Hm, why? Is the c-r envtest doing this as well? Why should it fail (I assume you want to tackle sporadic failures here)? 👀

Reply:
Yes, c-r is also doing this (see [here](https://github.com/kubernetes-sigs/controller-runtime/blob/10ae090c1d3ac0c560dfa1a29b2517eb8d74442b/pkg/envtest/server.go#L285)). That's why I added a `TODO` here.
It might be good to tackle some test flakes. But as we don't have any experience with running these tests in our pipelines yet, I just left itup to future PRs.

Reply:
Interesting, thanks for sharing the link. I think it's fine for now, let's keep it and get some experience 👍🏻